### DISCUSSION TITLE
Trouble Switching Audio Sources

### DISCUSSION THREAD

**Discussion by mrracoonman** [Rapptz/discord.py#7389] [Q&A]
I have a decent working discord music bot, but it runs off of pre-downloaded files. I like having a limited selection but I don't want to take up large amounts of space with the amount of files I have. Is there a way to change the audio source from a file location to a YouTube URL without downloading anything?

The code I'm currently working with is as follows: 
(I can never figure out the "insert code" thing here in Github, but it is in fact properly indented and has the proper imports. I have no idea why it looks the way it does, I am including lines to help clear thing up.)

________________________________________________________________________________________________________________________________________

@bot.command(aliases=["p", "P", "Play"])
async def play(ctx, *, arg=None):
    songData = songs[arg]
    audio_source = songData[0]
    x = songData[1]
    length = songData[2]
    image = songData[3]
    guild = ctx.guild
    voice_client: discord.VoiceClient = discord.utils.get(bot.voice_clients, guild=guild)

    if arg is not None:
        play = discord.Embed(title = "🦝 Playing! 🦝", color = discord.Color.dark_purple())
        play.add_field(name = f"{x} is {length} long.", value = "Starting now")
        play.set_image(url = image)
        print(f"{x} is now playing.")
        if not voice_client.is_playing():
            voice_client.play(audio_source, after=lambda e: print(f"{x} has finished playing."))
    elif arg is None:
        play = discord.Embed(title = "🦝 Can't Play That! 🦝", color = discord.Color.dark_purple())
        play.add_field(name = "Incorrect input. Please enter one of the options.", value = "If you do not know the options, type '^records' to see them.")

    await ctx.send(embed = play)

________________________________________________________________________________________________________________________________________

And my list of files looks somewhat like this: 

________________________________________________________________________________________________________________________________________

songs = _{"FIRST SONG IN LIST":_ _[discord.FFmpegPCMAudio(r'FILE LOCATION OF SAID MP3 FILE IN MY COMPUTER'),_ _"DISPLAY NAME FOR THE SONG IN DISCORD",_ _"SONG LENGTH",_ "_URL OF IMAGE I HAVE CHOSEN TO REPRESENT SONG IN DISCORD_"]}

________________________________________________________________________________________________________________________________________

Except it is over 50 songs long, with plans to extend it further, and that takes up way too much space.

To recap, I don't want to post a URL in the command when I use my bot, I want a selection from a list I have made. My problem is I haven't found anything that can let me change the audio sources from file locations to YouTube URLs.

I am grateful for any help, let me know if I can clear anything else up.
I am on Windows 10.

**Comment by ewanhowell5195** [comment]
You can use youtube_dl https://pypi.org/project/youtube_dl/

```python
from youtube_dl import YoutubeDL
with YoutubeDL({"format": "bestaudio", "quiet": True}) as ydl:
    info = ydl.extract_info("https://youtu.be/dQw4w9WgXcQ", download = False)
```
and then when you use FFmpegPCMAudio, enter `info["formats"][0]["url"]` instead of the file path

please note this is extremely simplified and will probably need a lot more to function correctly, but this should be the basics.

**Reply by mrracoonman** [reply]
What you seem to have provided only allowed for the use of one link, whereas I need to have 50+. I also assumed that YouTube link was a placeholder, considering I followed it before using it in my code, so I replaced it with my own. I do have a few idea to try with what you have provided, but please do get back to me on this if you know a way to have 50+ separate links.
Thank you.

**Reply by ewanhowell5195** [reply]
replace the link with a variable and just make the variable whatever link you want to play

**Reply by mrracoonman** [reply]
I believe I attempted that. My list is set up kinda like you saw above:

songs = {"FIRST SONG IN LIST": [discord.FFmpegPCMAudio(r'FILE LOCATION OF SAID MP3 FILE IN MY COMPUTER'), "DISPLAY NAME FOR THE SONG IN DISCORD", "SONG LENGTH", "URL OF IMAGE I HAVE CHOSEN TO REPRESENT SONG IN DISCORD"],
    "SECOND SONG IN LIST": [discord.FFmpegPCMAudio(r'FILE LOCATION OF SAID MP3 FILE IN MY COMPUTER'), "DISPLAY NAME FOR THE SONG IN DISCORD", "SONG LENGTH", "URL OF IMAGE I HAVE CHOSEN TO REPRESENT SONG IN DISCORD"]}

And so on for 50+ songs. in my original file, you can see it goes:

songData = songs[arg]
audio_source = songData[0]
x = songData[1]
length = songData[2]
image = songData[3]

to import the desired data dependent on the argument given by the user. I added a new one with the link and called it "y", while also providing a link within quotes before the discord.FFmepgPCMAudio, so it was ["LINK", discord.FFmepgPCMAudio(THE STUFF I ADDED THAT YOU PROVEDED), and then so on. . .
so in the end, the main file had:

songData = songs[arg]
y = songData[0]
audio_source = songData[1]
x = songData[2]
length = songData[3]
image = songData[4]

It kept giving an import error due to a circular import.

**Reply by ewanhowell5195** [reply]
add me on discord i can help you there easier ewanhowell5195#5195

**Reply by mrracoonman** [reply]
ok
